### PR TITLE
Support DeepGEMM for standard MoE dispatch

### PR DESCRIPTION
**Pull request by YAMY1234**
## Motivation

The `deep_gemm` MoE runner can be selected with the standard token dispatcher, but the existing integration assumes the DeepEP weight and activation layouts. Standard MoE layers therefore do not consistently requantize both expert weights to UE8M0, and the standard dispatch path does not provide a compact local-expert layout for DeepGEMM.

This change completes the standard-dispatch integration while preserving the existing DeepEP path.

Co-authored-by: Sam Li (@samuellees)

## Modifications

- Requantize both gate/up and down expert weights when a standard MoE layer uses the DeepGEMM runner.
- Quantize Blackwell FP8 activations directly with their UE8M0 scales instead of rounding the scales after quantization.
- Use the masked DeepGEMM layout when expert parallelism keeps the local expert buffer small.
- Use a compact, graph-static layout for the remaining standard-dispatch cases, including tensor-parallel execution without expert parallelism.
- Ignore non-local expert assignments during scatter and post-permute while preserving CUDA Graph-compatible buffer shapes.
- Use an uninitialized BF16 compact input buffer because scatter writes every valid row before the grouped matrix multiplication consumes it.
- Mark every compact-layout padding row with expert index `-1`, so DeepGEMM skips padding without a separate initialization kernel.
- Extend the existing UE8M0 and MoE scatter tests to cover standard-layer weight processing and real masked-versus-compact DeepGEMM execution.
- Extend the existing end-to-end standard-dispatch kernel test to cover FP8 and BF16 expert weights and validate every padding index.

## Tests

The existing registered MoE kernel test file completed with **52 passed** on an NVIDIA GB300. The FP8 and BF16 cases both execute the standard-dispatch pre-permute, grouped matrix multiplications, activation, down projection, and post-permute path.

## Accuracy Tests

Model-level validation completed at [commit `02d9f2e642`](https://github.com/sgl-project/sglang/commit/02d9f2e642907ceadb60279165a9359a47d436b4) with `Qwen/Qwen3.5-397B-A17B-FP8` on 16 NVIDIA GB300 GPUs. The server used PP2/TP8/EP1, the standard dispatcher with the DeepGEMM MoE runner, FP8 weights, BF16 KV and SSM caches, and full decode CUDA Graph through batch size 256. The configured breakable prefill graph was skipped by the existing runtime compatibility guard because some model layers do not apply Standard GQA.

```bash
python -m sglang.test.run_eval \
  --base-url http://localhost:8000 \
  --eval-name gsm8k \
  --num-examples 1319 \
  --num-shots 5 \
  --num-threads 256 \
  --max-tokens 16384 \
  --temperature 0
```

The evaluator completed every scored request without request-level errors: **1,288 correct and 26 incorrect, for 0.9802130898 accuracy (1,288/1,314)**. The 1,314 denominator is expected because the five-shot harness reserves five records from the 1,319-record GSM8K test file as fixed demonstrations and scores the remaining records.

## Speed Tests

The matched matrix ran at [commit `02d9f2e642`](https://github.com/sgl-project/sglang/commit/02d9f2e642907ceadb60279165a9359a47d436b4). The current head adds only [Black formatting](https://github.com/sgl-project/sglang/commit/bc7a37ab3f); an AST comparison excluding source positions is identical before and after that formatting commit.

Both lanes used the same Qwen3.5 checkpoint, 16 NVIDIA GB300 GPUs, PP2/TP8/EP1, exact 8,192-token input and 1,024-token output lengths, and concurrency 1–128. Each lane was repeated three times. The only runtime difference was:

```text
moe-runner-backend: flashinfer_trtllm -> deep_gemm
```

Values are median total tokens/s/GPU across the three runs. Classification compares the full three-run ranges; every range below is non-overlapping.

| Concurrency | FlashInfer TRTLLM | DeepGEMM | Change | Classification |
|---:|---:|---:|---:|:---|
| 1 | 75.49 | 55.18 | -26.90% | DeepGEMM slower |
| 2 | 104.87 | 77.79 | -25.82% | DeepGEMM slower |
| 4 | 207.41 | 157.64 | -23.99% | DeepGEMM slower |
| 8 | 398.99 | 283.63 | -28.91% | DeepGEMM slower |
| 16 | 697.80 | 484.15 | -30.62% | DeepGEMM slower |
| 32 | 1,140.88 | 743.62 | -34.82% | DeepGEMM slower |
| 64 | 1,715.61 | 1,037.06 | -39.55% | DeepGEMM slower |
| 128 | 2,420.24 | 1,410.13 | -41.74% | DeepGEMM slower |

All 48 formal results completed their full request counts with exact input and output lengths and no request-level errors. Full decode CUDA Graph captured on all 16 ranks in all six runs. The requested breakable prefill graph was skipped by the same model compatibility guard in both lanes. Runtime logs also confirm that every candidate rank selected the DeepGEMM runner and exercised grouped-contiguous DeepGEMM JIT.

This standard-dispatch implementation is functionally correct for the tested Qwen3.5 PP2/TP8/EP1 configuration, but it is not a performance win over FlashInfer TRTLLM at concurrency 1–128. The measured regression ranges from 23.99% to 41.74%, and no crossover appears in this range.

## Checklist

- [x] Format the code according to the [pre-commit guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [x] Add unit tests according to the [testing guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [x] Consider documentation according to the [documentation guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations). This change adds no user-facing option.
- [x] Provide accuracy and speed results according to the [accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [benchmark](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed) guidance.
- [x] Follow the SGLang [code-style guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, and `/rerun-failed-ci`.
4. After green CI and required approvals, ask Merge Oncalls or people with write permission to merge the pull request.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #30689441051](https://github.com/sgl-project/sglang/actions/runs/30689441051)
Latest PR Test (Extra): :white_check_mark: [Run #30689440964](https://github.com/sgl-project/sglang/actions/runs/30689440964)

### CODE REVIEW

File: python/sglang/srt/layers/moe/moe_runner/deep_gemm.py [hunk 1]
```diff
@@ -685,42 +707,166 @@ def pre_permute_standard_to_deep_gemm(
...
-            runner_config.num_local_experts,
-            hidden_states,
-            runner_config.top_k,
-            quant_info.block_shape,
-            output_dtype=output_dtype,
-            use_mxfp8=quant_info.use_mxfp8,
+    if output_dtype == torch.bfloat16:
+        packed_input_source = hidden_states
+        packed_input_source_scale = None
+        packed_input = torch.zeros(
```
Comment (ch-wan) [inline python/sglang/srt/layers/moe/moe_runner/deep_gemm.py]: can we use `torch.empty`?

Comment (YAMY1234) [reply python/sglang/srt/layers/moe/moe_runner/deep_gemm.py]: Yes. I changed the BF16 compact buffer to `torch.empty`. Scatter writes every valid row before the grouped matrix multiplication reads it, and padding rows now have expert index `-1`, so DeepGEMM skips them.

**Comment by YAMY1234** [comment]
/rerun-test test/registered/unit/hardware_backend/mlx/test_attn_dp_request_capacity.py

**Review by ch-wan — APPROVED** [review]

**Comment by hnyls2002** [comment]
FYI — I had a similar change in flight (#33342, now closed in favor of this one) and benchmarked the contiguous path against masked on two shapes that `_should_use_masked_standard_layout` also routes to contiguous:

- 128 local experts, no EP, topk 8 (block-fp8): 15-19% slower than masked for 512-4096 tokens, crossing over only at 8192.
- 48 local experts, tp8/ep8, topk 6 (fp4 experts): 22-30% slower across 512-8192.

Both have `num_local_experts > 32`. `_get_compact_all_tokens` lands within ~128 rows of the bound I measured against (both come out to roughly `tokens*topk + num_local*BLOCK_E`), so the numbers should carry over.

The cost driver is the per-expert padding `num_local * BLOCK_E`, which is largest exactly when EP is off or shallow — i.e. the configs this predicate sends down the contiguous path. Might be worth a sanity check on a low-EP shape.

One upside I did measure: at 16384 tokens the masked path OOMs trying to allocate its `[E, m_max, N]` buffer where contiguous completes fine.

@YAMY1234
